src/mpid/ch4/shm/ipc/src/ipc_win.c

```diff
@@ -0,0 +1,225 @@
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ipc_noinline.h"
+#include "ipc_mem.h"
+
+/* Return global node rank of each process in the shared communicator.
+ * I.e., rank in MPIR_Process.comm_world->node_comm. The caller routine
+ * must allocate/free each buffer. */
+static int get_node_ranks(MPIR_Comm * shm_comm_ptr, int *shm_ranks, int *node_ranks)
+{
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Group *shm_group_ptr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_GET_NODE_RANKS);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_GET_NODE_RANKS);
+
+    for (i = 0; i < shm_comm_ptr->local_size; i++)
+        shm_ranks[i] = i;
+
+    mpi_errno = MPIR_Comm_group_impl(shm_comm_ptr, &shm_group_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Get node group if it is not yet initialized */
+    if (!MPIDI_IPC_global.node_group_ptr) {
+        mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world->node_comm,
+                                         &MPIDI_IPC_global.node_group_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    mpi_errno = MPIR_Group_translate_ranks_impl(shm_group_ptr, shm_comm_ptr->local_size,
+                                                shm_ranks, MPIDI_IPC_global.node_group_ptr,
+                                                node_ranks);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Group_free_impl(shm_group_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_GET_NODE_RANKS);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
+typedef struct IPC_win_shared_info {
+    uint32_t disp_unit;
+    size_t size;
+    MPIDI_SHM_IPC_type_t ipc_type;
+    MPIDI_IPC_mem_handle_t mem_handle;
+} IPC_win_shared_info_t;
+
+int MPIDI_IPC_mpi_win_create_hook(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_IPC_win_t *ipc_win = NULL;
+    MPIR_Comm *shm_comm_ptr = win->comm_ptr->node_comm;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    int i;
+    size_t total_shm_size = 0;
+    MPIDIG_win_shared_info_t *shared_table = NULL;
+    IPC_win_shared_info_t *ipc_shared_table = NULL;     /* temporary exchange buffer */
+    int *ranks_in_shm_grp = NULL;
+    MPIDI_SHM_IPC_type_t ipc_type = MPIDI_SHM_IPC_TYPE__NONE;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_MPI_WIN_CREATE_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_MPI_WIN_CREATE_HOOK);
+    MPIR_CHKPMEM_DECL(2);
+    MPIR_CHKLMEM_DECL(2);
+
+    if (!shm_comm_ptr)
+        goto fn_exit;
+
+    /* Determine submodule availability and IPC type based on buffer type.
+     * If no submodule is available, skip initialization; if any submodule
+     * is on, we need always exchange in case remote buffer can be shared by IPC. */
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+    if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE != -1) {
+        ipc_type = MPIDI_SHM_IPC_TYPE__XPMEM;
+    } else {
+        goto fn_exit;
+    }
+#else
+    goto fn_exit;
+#endif
+
```

Comment:
I need to do more fix for SHM RMA to make it work correctly with GPU/host-mixed window buffers. For example, if only XPMEM is enabled, then we should allow SHM RMA for the host win buffers. Similarly, if only GPU is enabled, we should allow SHM RMA for GPU win buffers but use AM-based RMA for others.

The current SHM RMA is written in POSIX, and is enabled/disabled for the entire window (check `MPIDIG_WIN(win, shm_allocated)`) because we consider only host-buffer. We need make two changes:
(1) change `MPIDIG_WIN(win, shm_allocated)` to a per-process flag for host/GPU mixed case
(2) move the SHM RMA routines into SHM glue layer because it seems not right to put them in the POSIX submodule with the current IPC structure.

Reply:
I think we can delay mixed XPMEM/GPU modes for a separate PR.  For now, it's OK if only one or the other is picked at runtime.  For example, disabling XPMEM optimizations if GPU buffers are used.

Reply:
OK. We will then add a collective buffer type check in IPC_win_create in the GPU PR.